### PR TITLE
BUGFIX: SD-177 hide axis when min/max is null or out of range of data or min > max

### PR DESCRIPTION
**Pull request by naibinh**

### CODE REVIEW

**Comment by naibinh** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1019/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2522/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/598/ : UNSTABLE in 20m 05s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2525/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1919/ : SUCCESS in 2m 03s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2526/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1921/ : SUCCESS in 1m 44s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2527/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1922/ : SUCCESS in 2m 41s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2529/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1923/ : SUCCESS in 1m 46s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1024/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2530/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/602/ : SUCCESS in 15m 00s

File: src/components/visualizations/chart/highcharts/plugins/test/getTickPositions.spec.ts [hunk 1]
```diff
@@ -365,4 +372,193 @@ describe('get tick positions for dual axes chart', () => {
             });
         });
     });
+
+    describe('test generate tick positions with highchart', () => {
+        const AXIS_PADDING = 0.05;
+        let getTickPositions: Function;
+
+        const getAxis = (index = 0,
+                         dataMin = undefined as number, dataMax= undefined as number,
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/plugins/test/getTickPositions.spec.ts]: whi you are using 'as'?  Why not simple 'dataMin: number = undefined'?

File: src/components/visualizations/chart/highcharts/plugins/zeroAlignYAxis.ts [hunk 2]
```diff
@@ -246,56 +246,60 @@ export function getTickAmount(tickPosition: number[]): number {
...
-                        max: axis.max,
-                        isSetMin: axis.userOptions.min !== undefined,
-                        isSetMax: axis.userOptions.max !== undefined
-                    });
+        const yAxes = chart.axes.filter((axis: any) => axis.coll === 'yAxis');
+
+        for (const yAxisIndex in yAxes) {
+            if (yAxes.hasOwnProperty(yAxisIndex)) {
+                const axis = yAxes[yAxisIndex];
+                if (typeof axis.max === 'undefined' || axis.max === null ||
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/plugins/zeroAlignYAxis.ts]: similar check as above, can be in helper function

File: src/components/visualizations/chart/highcharts/plugins/zeroAlignYAxis.ts [hunk 3]
```diff
@@ -246,56 +246,60 @@ export function getTickAmount(tickPosition: number[]): number {
...
-                        id: axis.options.index,
-                        min: axis.min,
-                        max: axis.max,
-                        isSetMin: axis.userOptions.min !== undefined,
-                        isSetMax: axis.userOptions.max !== undefined
-                    });
+        const yAxes = chart.axes.filter((axis: any) => axis.coll === 'yAxis');
+
+        for (const yAxisIndex in yAxes) {
+            if (yAxes.hasOwnProperty(yAxisIndex)) {
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/plugins/zeroAlignYAxis.ts]: this processing of other axes can be moved to separate function

Comment (naibinh) [reply src/components/visualizations/chart/highcharts/plugins/zeroAlignYAxis.ts]: Actually, `for...in` and `hasOwnProperty` must be a pair to bypass typescript check.
It does not really relate to process opposite axis.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2537/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1932/ : SUCCESS in 6m 29s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2538/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1933/ : SUCCESS in 3m 02s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1027/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2539/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/603/ : UNSTABLE in 20m 03s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2540/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1934/ : SUCCESS in 4m 18s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1028/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2541/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/604/ : SUCCESS in 25m 21s

**Comment by naibinh** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2542/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1935/ : FAILURE in 1m 17s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1029/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2543/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/605/ : SUCCESS in 17m 42s

**Comment by pbenes** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1052/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2588/ : SUCCESS in 16s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/616/ : UNSTABLE in 23m 50s

File: stories/internal/DualAxesChartConfig.tsx [hunk 4]
```diff
@@ -90,7 +93,31 @@ storiesOf('Internal/DualAxesMinMaxConfig', module)
             </div>
         );
     })
-    .add('Dataset with 0+ data on both axes, min >= max', () => {
+    .add('Dataset with 0+ data on both axes, min > max on left axis', () => {
```
Comment (pbenes) [inline stories/internal/DualAxesChartConfig.tsx]: I see that there are many removed screenshots from stories (14) and only a few added (4). Could you please check and explain why these are no longer needed? Thanks :-)

**Comment by naibinh** [comment]
Question:
I have not tried myself, but what about something like this?

```yAxes.forEach((axis, yAxisIndex) => {
    generateAxisMinMax(minmax, axis, yAxisIndex, isLineChartType);
```});

I cannot comment above so I move it here.
The point is that we cannot break the loop by using `forEach`.
There is an `if` that stops the loop.

File: src/components/visualizations/chart/highcharts/plugins/zeroAlignYAxis.ts [hunk 5]
```diff
@@ -35,6 +35,14 @@ export interface IMinMaxLookup {
     1?: IMinMaxInfo;
 }
 
+/**
+ * Check value is either null or undefined
+ * @param value
+ */
+function isNotDefined(value: number) {
+    return value == null; // this assert applies to null and undefined
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/plugins/zeroAlignYAxis.ts]: done

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2594/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/1988/ : SUCCESS in 1m 58s

File: stories/internal/DualAxesChartConfig.tsx (hunk 4 above)
Comment (pbenes) [reply stories/internal/DualAxesChartConfig.tsx]: :+1:

**Comment by pbenes** [comment]
@naibinh  thanks for explanation, I did not notice the `return` there at first. I think we can squash the commits and let me know, I can add merge label then.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/2607/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2002/ : SUCCESS in 2m 09s

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/566/ : SUCCESS in 22s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2003/ : SUCCESS in 2m 40s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1063/ : UNSTABLE in 16m 07s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/532/ : SUCCESS in 17m 40s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/567/ : SUCCESS in 18s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/2004/ : SUCCESS in 7m 09s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1064/ : SUCCESS in 17m 14s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/533/ : SUCCESS in 5m 23s

**Pull request merged**